overlord/devicestate/handlers.go

```diff
@@ -63,13 +71,60 @@ func deviceAPIBaseURL() string {
 }
 
 var (
-	keyLength        = 4096
-	retryInterval    = 60 * time.Second
-	deviceAPIBase    = deviceAPIBaseURL()
-	requestIDURL     = deviceAPIBase + "request-id"
-	serialRequestURL = deviceAPIBase + "devices"
+	keyLength        uint64 = 4096
+	retryInterval           = 60 * time.Second
+	deviceAPIBase           = deviceAPIBaseURL()
+	requestIDURL            = deviceAPIBase + "request-id"
+	serialRequestURL        = deviceAPIBase + "devices"
+	sshKeyFile              = "/tmp/snapd.key.tmp"
+	sshPublicKeyFile        = sshKeyFile + ".pub"
 )
 
+func rsaGenerateKey() (*rsa.PrivateKey, error) {
+	defer func() {
+		os.Remove(sshKeyFile)
+		os.Remove(sshPublicKeyFile)
+	}()
+
+	os.Remove(sshKeyFile)
+	os.Remove(sshPublicKeyFile)
+
+	switch C.rsa_generate_key(C.uint64_t(keyLength), C.CString(sshKeyFile), C.CString(sshPublicKeyFile)) {
```

Comment:
Mainly because building up an rsa.PrivateKey is non-trivial and I'm using the parsing functions to get the job done. I haven't looked too deeply if I could leverage the openssl PEM encoding functions to place the encoded key into memory.